feature-pitches/002-file-based-workflows/002-file-based-workflows.md

```diff
@@ -0,0 +1,109 @@
+
+Additionally, there is a common challenge we run into when introducing Renku workflows to users who *do* want to use workflows: it is cumbersome/just-not-tenable to create and develop (modify) workflows of more than a few steps. Many workflow users have 10+ steps in their workflows [[source](https://sdsc.atlassian.net/wiki/spaces/RENKU/pages/2246541341/Reproducibility+on+HPC)]. Entering 10+ steps on the command line is tedious to do once, but even more significant is that modifying that workflow- which users do all the time as they develop their pipeline- requires recreating and rewiring steps again and again on the CLI. Some users end up writing a script to call a series of Renku run commands [[ORDES example](https://renkulab.io/projects/renku-stories/digirhythm/files/blob/src/build_workflow.sh)]. Renku should provide a built-in way to create and develop multi-step workflows via a file!
+
+## 🍴 Appetite
+
+We’d like to invest 2 sprints (6 weeks) for making workflows more accessible to a broader audience.
+
+## 🎯 Solution
+
+We’d like to offer the ability to define Renku workflows in a file, where the full workflow is laid out for the user to plainly see and edit.
+
+I’ve done a [survey of popular workflow tools](https://sdsc.atlassian.net/wiki/spaces/RENKU/pages/2310701105/Workflow+Tool+Research), and YAML seems to be a common choice for a workflow file, but we leave the final decision to the team.
+
+<img src="fat-marker-file.png" width="400">
+
+The workflow file names a sequence of steps, and each step names its command, inputs, outputs, parameters, etc. Each step should be named so that it can be referenced for running individually. I imagine that the structure of the workflow step definition in the file would be pretty similar to the output of `renku workflow show`.
+
+![renku workflow show](workflow-show.png)
+
+For the scope of this pitch, expressing `renku workflow iterate` in the workflow file definition syntax is _out of scope_, along with any kinds of looping or branching.
+
+### Editing workflow files
+
+The workflow file is opened and modified just like any other code file in the project. This pitch requires no changes to the UI.
+
+The user should be able to create more than one workflow file in a project.
+
+### Running workflow files
+
+The user would still use the CLI to run the workflow, referencing the workflow via the workflow file name. To improve usability for non-CLI users, in tutorials, we could perhaps demonstrate running shell commands from inside a notebook (`!renku workflow execute workflow.yml`).
+
+Developing a workflow should be as simple to the user as editing the workflow file, executing the workflow file on the CLI, making further modifications to the workflow file, executing the workflow file again, etc. Renku should not interrupt or slow down this development process. 
```

Comment:
Yea, I agree that the need to commit before every renku run/workflow command is a small but constant annoyance to usability. What do you mean by "dry-run" exactly? Would it just check that the workflow file syntax is legitimate? 

One option to consider- though I recognize that this is opening a fairly new direction for Renku- is to have `renku workflow execute` automatically do a commit before executing the workflow, so the user doesn't have to do that extra step every time. I haven't thought through this idea very much... But we could offer something like a `renku workflow execute <workflow> --commit-for-me` flag that defaults `True`, but the user could turn off if they don't want commits happening for them...? Totally not sure about this idea, just throwing it out there. I agree the extra commit step is annoying to users. I've even seen it stumble up Renku team members when they're just trying to demo something.

Reply:
right, I think "dry-run" was bad naming - I meant more like just running the workflow as prescribed, but do not commit anything. There was something like this already being done because we need to delay committing in the case that the execution is happening remotely. The idea we had in the past was something like a renku "transaction" - you could do a bunch of work using the Renku CLI and Renku would keep track of the metadata somewhere, but wouldn't commit anything until you told it to.

Reply:
This feels like it's opening a whole separate topic that could be it's own pitch...?

Reply:
I think it will likely come up during design/testing/ux validation so maybe worth adding it to the out-of-scope section.

Reply:
Added.